brainiak/utils/fmrisim.py

```diff
@@ -1595,6 +1603,66 @@ def _generate_noise_temporal_drift(trs,
         phase = (timepoints / (trs - 1) * cycles * 2 * np.pi) + phaseshift
         noise_drift = np.sin(phase)
 
+    elif basis == 'cos_power_drop':
+
+        # Make a vector counting each TR
+        timepoints = np.linspace(0, trs - 1, trs) * tr_duration
+
+        # Specify the other timing information
+        duration = trs * tr_duration
+
+        # How bases do you have?
+        basis_funcs = int(np.floor(2 * duration))
+
+        noise_drift = np.zeros((timepoints.shape[0], basis_funcs))
+        for basis_counter in list(range(1, basis_funcs + 1)):
+            # What steps do you want to take for this basis function
+            random_phase = np.random.rand() * np.pi * 2
+
+            timepoint_phase = (timepoints / duration * np.pi * basis_counter)
```

Comment:
Here, `basis_counter` goes from 1 to `basis_funs`, which is twice the total duration of the scan, in the unit of a second. `timepoints` is equally spaced time points between 0 and `duration`. When it is divided by `duration`, the result is equally spaced numbers between 0 and 1. This times pi would result in equally spaced points between 0 and pi. But multiply this phase with `basis_counter` which is a duration seems a bit strange, to be. It won't mean something as a radius anymore, but would be added with `random_phase` and be used as an input to cosine function. I think the unit here may be somehow wrong. Likely, the `basis_funcs` should be something like the ratio between the lowest frequency of the fft of the time series to the Nyquist frequency (twice of the sampling frequency of the scanner) instead, in line 1615?

Reply:
Thanks. Now I think I may have got it. You are trying to make a power spectrum that decreases exponentially from the lowest frequency (inverse of double the duration of the scan, corresponding to when `basis_counter` equals 1) to the highest frequency (the sampling frequency of fMRI,  when `basis_counter` equals `basis_funcs`).

Reply:
But then I am confused with the description "the power drop is meant to drop off to 99% by basis_funcs at duration, with the remaining power being captured in the subsequent basis_funcs duration->2*duration." Can I interpret it as, if I am integrating an exponential decaying curve from 0 up to certain number (duration) here, the integral is 99% of the integral from 0 to 2*duration? If this is the case, I would imagine the power of the numerator and denominator in the function `power_drop` should differ in a ratio of 2, instead of a ratio of `F`?

Reply:
Actually I think my second to last comment was incorrect. The highest frequency considered here is not the sampling frequency of fMRI, but a frequency which is set to the value of `duration`. This means that if our scan is 300 s long, we are talking about the highest frequency of 300 Hz (to see this, the total phase change of `timepoint_phase` from its first element to the last element is `2 * duration * pi` when `basis_counter` is `basis_funcs`. Theoretically there seems to be no restriction of what we set as the highest frequency to consider. But why do we make this choice? We never deal with 300 Hz signals in fMRI anyway. It seems to me the more natural choice is to set the highest frequency as the Nyquest frequency (2/TR). In this case, `basis_funcs` can be set as `duration / TR`.

Reply:
This is a really tricky point so I will do my best to work through it but I am not sure if I can give a satisfying answer. Firstly, I think you implied that `r` is affected by the choice of what `basis_funcs` is; this is not the case, only `duration` and `period` affect `r`. Hence how many `basis_funcs` we estimate will ultimately be non-influential to the `r`. Secondly, pragmatically, this choice doesn't matter because the drop off is such that the influence of any basis function after about the 10th is close to zero. For reference, for a typical `r`, the 10th function has a weight of 2e-2, 50th weight is 4e-9, 100th is 2e-17. So the difference between going to 300 or 600 is negligible. Thirdly, that said, it isn't so much about what practically matters but what is theoretically correct. The way I think we were thinking of it is that we wanted to compensate for Nyquist by making sure we had sufficient resolution at the frequency of the duration. Does this seem reasonable?

Reply:
Yes. But if it were not the case that `basis_funcs` was set to `2 * duration`, I don't understand what the meaing of `4 * L` is as the power in the denominator. When I say `basis_funcs` infuence `r`, I mean that you take a frequency which equals `duration` as your upper frequency, and it is this frequency that influences `r`. I think I am mainly concerned that such a choice of the highest frequency does not make sense.

However, putting that aside, you are right that in most cases, when you use `4 * L` as the power (regardless of what meaning it is), your denominator would essentially be 1. And in fact, if you consider 99% as the target of power you retain up to the frequency corresponding to `period` out of the total power up to the frequency of *infinity*, then you can keep the definition of the `numerator` but let the `denominator` be 1. This way, you don't really need to do any fitting. 
You can directly solve the equation of `numerator = 0.99` to get `r`. There is also no more confusion of what needs to be in the `denominator`, and in this case I also agree that `basis_funcs` does not influence `r`.

Reply:
@lcnature I was convinced of your argument by this last message and so reached out to Chris and he agreed. A small tweak he suggested, that I think you also thought of, is to base the maximum frequency on the number of TRs, not the duration of the run. He gives the example that if you had a 100s TR for a 200s run you wouldn't want to to simulate 200 basis funcs but instead only 2. The code change I made should accord with that. Moreover, I made a change to the `power_drop` function in order to deal with the new `basis_funcs`. However, I kept the denominator as a variable rather than 1. Although practically for most cases you are right that it will be equal to 1, for short runs (e.g. <100s) this value will be non-trivially below 1. Plus, including that optimization is a quick step done only once per simulation.

Reply:
Thanks @CameronTEllis . It is better now. But I still think that if you don't change the denominator to 1, the power in the denominator would be something like `2 * trs` - the power should be unit-less (I know practically they both make the denominator so close to 1 with almost no difference). I was almost convinced that the 4 in the original code makes sense but I am uncertain now. @cbaldassano thoughts? Should it be `numerator = 1 - r ** (4 * L / F)` for `2 * (1/F / 1/2L)` and `denominator = 1 - r ** (2 * trs)` for `2 * 1/2tr / 1/2L`, which is `2 * trs`? Here, `1/F` is the frequency of "retaining", `1/2tr` is the Nyquist frequency and `1/2L` is the lowest non-zero frequency in the spectrum? I am a bit uncertain whether it should be `/1L` or `1/2L`.

Reply:
@lcnature  is right that the units aren't right there. The reasoning for having the sum be up to L isn't based on the Nyquist freq (though it is related), it is because that is the number of basis functions in the DCT. For example see the top of [pg 2 here](https://link.springer.com/content/pdf/10.1007%2F978-1-4471-6684-9_20.pdf) - u indexes the timepoints, m indexes the basis functions. The twos are coming from the geometric series equation, since power is proportional to r**2 (so these terms could be written, for example, (r**2)^(L/F)).

I believe the numerator is correct as is but the denominator needs to be put back into # of TRs units, e.g.

```
            numerator = 1 - r ** (2 * L / F)  # Power of this period
            denominator = 1 - r ** (2 * L / tr_duration)  # Power of all periods
```

Reply:
Thanks for this clarification @cbaldassano , it helped my understanding. Because of it I added, these changes, some documentation, an example and a check to ensure that `tr_duration` < `period`. However, reading this did raise a concern about another line of code https://github.com/brainiak/brainiak/blob/e21342d26764f245765378e59edb5f3c90985d40/brainiak/utils/fmrisim.py#L1679 (`basis_weights = r ** np.arange(basis_funcs)`). This line is meant to find weights for each of the basis functions before they are all added together; however, based on my new understanding shouldn't this be: `basis_weights = r ** (2 * np.arange(basis_funcs))` in order to make power proportional?

Reply:
No, the power spectrum is the square of the coefficients. So the coefficients themselves are 1, r, r^2, r^3... and the power spectrum of the frequencies is 1, r^2, r^4, r^6...

Reply:
Aha, okay thanks for that clarification. The code should then be ready to go, what do you think @lcnature ?

Reply:
Looks great. I am glad to approve it!